<a href="https://colab.research.google.com/github/ananyapoonsawat/BSC_DPDM23/blob/main/Homework_1_group.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [2]:
import pandas as pd
import os

In [3]:
from google.colab import drive
drive.mount('/content/drive')
path = '/content/drive/My Drive/bsc_dpdm23_data'

Mounted at /content/drive


In [5]:
data = pd.read_csv(os.path.join(path,'/content/drive/MyDrive/BSC_DPDM_23/BSC_dpdm23_data/reduced_marketbasket.csv'),encoding='latin-1')

In [6]:
data

InvoiceNo StockCode                          Description  Quantity  \
0         536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1         536365     71053                  WHITE METAL LANTERN         6   
2         536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3         536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4         536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   
...          ...       ...                                  ...       ...   
541904    581587     22613          PACK OF 20 SPACEBOY NAPKINS        12   
541905    581587     22899         CHILDREN'S APRON DOLLY GIRL          6   
541906    581587     23254        CHILDRENS CUTLERY DOLLY GIRL          4   
541907    581587     23255      CHILDRENS CUTLERY CIRCUS PARADE         4   
541908    581587     22138        BAKING SET 9 PIECE RETROSPOT          3   

             InvoiceDate  UnitPrice  CustomerID         Country  
0       12/01/2010 08:26       2.55     17850.0  United Kingdom  
1       12/01/2010 08:26       3.39     17850.0  United Kingdom  
2       12/01/2010 08:26       2.75     17850.0  United Kingdom  
3       12/01/2010 08:26       3.39     17850.0  United Kingdom  
4       12/01/2010 08:26       3.39     17850.0  United Kingdom  
...                  ...        ...         ...             ...  
541904  12/09/2011 12:50       0.85     12680.0          France  
541905  12/09/2011 12:50       2.10     12680.0          France  
541906  12/09/2011 12:50       4.15     12680.0          France  
541907  12/09/2011 12:50       4.15     12680.0          France  
541908  12/09/2011 12:50       4.95     12680.0          France  

[541909 rows x 8 columns]

####เคลียร์ข้อมูล โดยต้องการ เดือน/วัน/ปี จาก Column InvoiceDate

In [10]:
data [ 'InvoiceDate' ] [ 0 ] .split()[0] #แยกวันที่กับเวลา

'12/01/2010'

In [11]:
date = []
for i in range(len(data['InvoiceDate'])):
  x = data['InvoiceDate'][i].split()[0]
  date.append(x)
#set(L)
date[0].split('/') #จะได้

['12', '01', '2010']

In [12]:
day =[] #แยกเดือนวันปีออกจากกัน
month =[]
year =[]
for i in range(len(data['InvoiceDate'])):
  x = date[i].split('/')[0]
  y = date[i].split('/')[1]
  z = date[i].split('/')[2]
  month.append(x)
  day.append(y)
  year.append(z)

In [14]:
data_UNset = data #data_UNsetคือข้อมูลที่ยังเซทไม่เสร็จ
data_UNset['date'] = date
data_UNset['day'] = day
data_UNset['month'] = month
data_UNset['year'] = year

In [15]:
set(data_UNset['date'])

{'01/04/2011',
 '01/05/2011',
 '01/06/2011',
 '01/07/2011',
 '01/09/2011',
 '01/10/2011',
 '01/11/2011',
 '01/12/2011',
 '02/01/2011',
 '02/02/2011',
 '02/03/2011',
 '02/04/2011',
 '02/06/2011',
 '02/07/2011',
 '02/08/2011',
 '02/09/2011',
 '02/10/2011',
 '02/11/2011',
 '03/01/2011',
 '03/02/2011',
 '03/03/2011',
 '03/04/2011',
 '03/06/2011',
 '03/07/2011',
 '03/08/2011',
 '03/09/2011',
 '03/10/2011',
 '03/11/2011',
 '04/01/2011',
 '04/03/2011',
 '04/04/2011',
 '04/05/2011',
 '04/06/2011',
 '04/07/2011',
 '04/08/2011',
 '04/10/2011',
 '04/11/2011',
 '04/12/2011',
 '05/01/2011',
 '05/03/2011',
 '05/04/2011',
 '05/05/2011',
 '05/06/2011',
 '05/08/2011',
 '05/09/2011',
 '05/10/2011',
 '05/11/2011',
 '05/12/2011',
 '06/01/2011',
 '06/02/2011',
 '06/03/2011',
 '06/05/2011',
 '06/06/2011',
 '06/07/2011',
 '06/08/2011',
 '06/09/2011',
 '06/10/2011',
 '06/12/2011',
 '07/01/2011',
 '07/03/2011',
 '07/04/2011',
 '07/05/2011',
 '07/06/2011',
 '07/07/2011',
 '07/08/2011',
 '07/10/2011',
 '07/11/20

In [16]:
set(data_UNset['day'])

{'01',
 '02',
 '03',
 '04',
 '05',
 '06',
 '07',
 '08',
 '09',
 '10',
 '11',
 '12',
 '13',
 '14',
 '15',
 '16',
 '17',
 '18',
 '19',
 '20',
 '21',
 '22',
 '23',
 '24',
 '25',
 '26',
 '27',
 '28',
 '29',
 '30',
 '31'}

In [17]:
x=[] #แปลงวัน
for i in range(1,32):
  x.append(i)
print(x)

[1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31]


In [18]:
data_UNset1 = data_UNset.replace(['01', '02', '03', '04', '05', '06', '07', '08', '09', '10','11',
                                         '12', '13', '14', '15', '16', '17', '18', '19', '20', '21', '22',
                                         '23', '24', '25', '26', '27', '28', '29', '30', '31'],x)

In [19]:
data_UNset.head()

InvoiceNo StockCode                          Description  Quantity  \
0    536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1    536365     71053                  WHITE METAL LANTERN         6   
2    536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3    536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4    536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   

        InvoiceDate  UnitPrice  CustomerID         Country        date day  \
0  12/01/2010 08:26       2.55     17850.0  United Kingdom  12/01/2010  01   
1  12/01/2010 08:26       3.39     17850.0  United Kingdom  12/01/2010  01   
2  12/01/2010 08:26       2.75     17850.0  United Kingdom  12/01/2010  01   
3  12/01/2010 08:26       3.39     17850.0  United Kingdom  12/01/2010  01   
4  12/01/2010 08:26       3.39     17850.0  United Kingdom  12/01/2010  01   

  month  year  
0    12  2010  
1    12  2010  
2    12  2010  
3    12  2010  
4    12  2010

In [20]:
data_UNset1.head()

InvoiceNo StockCode                          Description  Quantity  \
0    536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1    536365     71053                  WHITE METAL LANTERN         6   
2    536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3    536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4    536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   

        InvoiceDate  UnitPrice  CustomerID         Country        date  day  \
0  12/01/2010 08:26       2.55     17850.0  United Kingdom  12/01/2010    1   
1  12/01/2010 08:26       3.39     17850.0  United Kingdom  12/01/2010    1   
2  12/01/2010 08:26       2.75     17850.0  United Kingdom  12/01/2010    1   
3  12/01/2010 08:26       3.39     17850.0  United Kingdom  12/01/2010    1   
4  12/01/2010 08:26       3.39     17850.0  United Kingdom  12/01/2010    1   

  month  year  
0    12  2010  
1    12  2010  
2    12  2010  
3    12  2010  
4    12  2010

In [21]:
set(data_UNset1['month']) #แปลงเดือน

{1,
 '1',
 10,
 11,
 12,
 2,
 '2',
 3,
 '3',
 4,
 '4',
 5,
 '5',
 6,
 '6',
 7,
 '7',
 8,
 '8',
 9,
 '9'}

In [22]:
data_UNset2 = data_UNset1.replace(to_replace=["1", "2",'3','4','5','6','7','8','9'],value=[1,2,3,4,5,6,7,8,9])

In [24]:
data_UNset2.isnull().any()

InvoiceNo      False
StockCode      False
Description     True
Quantity       False
InvoiceDate    False
UnitPrice      False
CustomerID      True
Country        False
date           False
day            False
month          False
year           False
dtype: bool

In [25]:
data_UNset[data_UNset['month']=='01']

InvoiceNo StockCode                Description  Quantity  \
42481    539993     22386    JUMBO BAG PINK POLKADOT        10   
42482    539993     21499         BLUE POLKADOT WRAP        25   
42483    539993     21498        RED RETROSPOT WRAP         25   
42484    539993     22379   RECYCLING BAG RETROSPOT          5   
42485    539993     20718  RED RETROSPOT SHOPPER BAG        10   
...         ...       ...                        ...       ...   
55385    540992     21912   VINTAGE SNAKES & LADDERS         2   
55386    540992     51008   AFGHAN SLIPPER SOCK PAIR        10   
55387    540992     21508    VINTAGE KID DOLLY CARD         12   
55388    540992     22716         CARD CIRCUS PARADE        12   
55389    540992     22983        CARD BILLBOARD FONT        12   

            InvoiceDate  UnitPrice  CustomerID         Country        date  \
42481  01/04/2011 10:00       1.95     13313.0  United Kingdom  01/04/2011   
42482  01/04/2011 10:00       0.42     13313.0  United Kingdom  01/04/2011   
42483  01/04/2011 10:00       0.42     13313.0  United Kingdom  01/04/2011   
42484  01/04/2011 10:00       2.10     13313.0  United Kingdom  01/04/2011   
42485  01/04/2011 10:00       1.25     13313.0  United Kingdom  01/04/2011   
...                 ...        ...         ...             ...         ...   
55385  01/12/2011 16:57       3.75     17364.0  United Kingdom  01/12/2011   
55386  01/12/2011 16:57       3.45     17364.0  United Kingdom  01/12/2011   
55387  01/12/2011 16:57       0.42     17364.0  United Kingdom  01/12/2011   
55388  01/12/2011 16:57       0.42     17364.0  United Kingdom  01/12/2011   
55389  01/12/2011 16:57       0.42     17364.0  United Kingdom  01/12/2011   

      day month  year  
42481  04    01  2011  
42482  04    01  2011  
42483  04    01  2011  
42484  04    01  2011  
42485  04    01  2011  
...    ..   ...   ...  
55385  12    01  2011  
55386  12    01  2011  
55387  12    01  2011  
55388  12    01  2011  
55389  12    01  2011  

[12909 rows x 12 columns]

In [26]:
data_UNset2[data_UNset2['month']==1]

InvoiceNo StockCode                    Description  Quantity  \
42481    539993     22386        JUMBO BAG PINK POLKADOT        10   
42482    539993     21499             BLUE POLKADOT WRAP        25   
42483    539993     21498            RED RETROSPOT WRAP         25   
42484    539993     22379       RECYCLING BAG RETROSPOT          5   
42485    539993     20718      RED RETROSPOT SHOPPER BAG        10   
...         ...       ...                            ...       ...   
77623    542745     22138  BAKING SET 9 PIECE RETROSPOT         13   
77624    542746     21034   REX CASH+CARRY JUMBO SHOPPER         2   
77625   C542748     22969   HOMEMADE JAM SCENTED CANDLES       -12   
77626   C542753     22649   STRAWBERRY FAIRY CAKE TEAPOT        -1   
77627    542759     22649   STRAWBERRY FAIRY CAKE TEAPOT         1   

            InvoiceDate  UnitPrice  CustomerID         Country        date  \
42481  01/04/2011 10:00       1.95     13313.0  United Kingdom  01/04/2011   
42482  01/04/2011 10:00       0.42     13313.0  United Kingdom  01/04/2011   
42483  01/04/2011 10:00       0.42     13313.0  United Kingdom  01/04/2011   
42484  01/04/2011 10:00       2.10     13313.0  United Kingdom  01/04/2011   
42485  01/04/2011 10:00       1.25     13313.0  United Kingdom  01/04/2011   
...                 ...        ...         ...             ...         ...   
77623   1/31/2011 16:47       4.95     15881.0  United Kingdom   1/31/2011   
77624   1/31/2011 16:48       0.95     15881.0  United Kingdom   1/31/2011   
77625   1/31/2011 16:53       1.45     15881.0  United Kingdom   1/31/2011   
77626   1/31/2011 16:57       4.95     15881.0  United Kingdom   1/31/2011   
77627   1/31/2011 16:59       4.95     15881.0  United Kingdom   1/31/2011   

       day  month  year  
42481    4      1  2011  
42482    4      1  2011  
42483    4      1  2011  
42484    4      1  2011  
42485    4      1  2011  
...    ...    ...   ...  
77623   31      1  2011  
77624   31      1  2011  
77625   31      1  2011  
77626   31      1  2011  
77627   31      1  2011  

[35147 rows x 12 columns]

In [27]:
set(data['Country']) #แยกทวีปจากต้นทาง

{'Australia',
 'Austria',
 'Bahrain',
 'Belgium',
 'Brazil',
 'Canada',
 'Channel Islands',
 'Cyprus',
 'Czech Republic',
 'Denmark',
 'EIRE',
 'European Community',
 'Finland',
 'France',
 'Germany',
 'Greece',
 'Hong Kong',
 'Iceland',
 'Israel',
 'Italy',
 'Japan',
 'Lebanon',
 'Lithuania',
 'Malta',
 'Netherlands',
 'Norway',
 'Poland',
 'Portugal',
 'RSA',
 'Saudi Arabia',
 'Singapore',
 'Spain',
 'Sweden',
 'Switzerland',
 'USA',
 'United Arab Emirates',
 'United Kingdom',
 'Unspecified'}

In [28]:
#ทวีปยุโรป
EU_country = ['Austria','Belgium','Cyprus','Channel Islands','Czech Republic','Denmark','EIRE','European Community',
              'Finland','France','Germany','Greece','Iceland','Italy','Lithuania','Malta','Netherlands',
              'Norway','Poland','Portugal','Spain','Sweden','Switzerland','United Kingdom']

In [29]:
#ทวีปเอเชีย
Asia_country = ['Hong Kong','Bahrain','Israel','Japan','Lebanon','Saudi Arabia','Singapore','United Arab Emirates']

In [31]:
def finditem(x,y) :
  L=set(data[(data['StockCode']==x)|(data['StockCode']==y)]['Description'])
  return L ## Code สำหรับหาชื่อสินค้า จาก Column "Description"

## Homework 1 ส่งก่อนวันจันทร์หน้าๆนู้น
หา Frequent Item Set และ Association Rules

* ของช่วงเวลา Christmas & New Year ( 1 Dec - 31 Dec )
* ของช่วงเวลาในทวีป Europe
## ของประเทศในทวีป ASIA
* ของช่วงเวลา ตรุษจีน ในทวีป Asis ( 15 Jan - 15 Feb )
* ของช่วงเวลา Christmas & New Year ในทวีป Europe ( 1 Dec - 31 Dec )
* ของวันวาเลนไทน์ 14 Feb
* โดยจัดกลุ่มตามประเทศ ( ฟิลว่าในประเทศเดียวกัน เขาซื้ออะไรพร้อมๆกัน )

พร้อมแปลผล

พรีเซนท์ว่าได้ Item อะไรออกมา แล้วแต่ละ Item ทำอะไร เช่น Asia คนมักจะซื้ออะไร 2 อย่าง, 3 อย่าง ( Groupby แบบประเทศเดือน )

###1. ของช่วงเวลา Christmas & New Year ( 1 Dec - 31 Dec )

In [33]:
data_ChristmasNewYear = data_UNset2[data_UNset2['month']==12] #data_UNset2คือเซทที่เคลียร์ล่สุดแล้ว
data_ChristmasNewYear

InvoiceNo StockCode                          Description  Quantity  \
0         536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1         536365     71053                  WHITE METAL LANTERN         6   
2         536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3         536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4         536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   
...          ...       ...                                  ...       ...   
541904    581587     22613          PACK OF 20 SPACEBOY NAPKINS        12   
541905    581587     22899         CHILDREN'S APRON DOLLY GIRL          6   
541906    581587     23254        CHILDRENS CUTLERY DOLLY GIRL          4   
541907    581587     23255      CHILDRENS CUTLERY CIRCUS PARADE         4   
541908    581587     22138        BAKING SET 9 PIECE RETROSPOT          3   

             InvoiceDate  UnitPrice  CustomerID         Country        date  \
0       12/01/2010 08:26       2.55     17850.0  United Kingdom  12/01/2010   
1       12/01/2010 08:26       3.39     17850.0  United Kingdom  12/01/2010   
2       12/01/2010 08:26       2.75     17850.0  United Kingdom  12/01/2010   
3       12/01/2010 08:26       3.39     17850.0  United Kingdom  12/01/2010   
4       12/01/2010 08:26       3.39     17850.0  United Kingdom  12/01/2010   
...                  ...        ...         ...             ...         ...   
541904  12/09/2011 12:50       0.85     12680.0          France  12/09/2011   
541905  12/09/2011 12:50       2.10     12680.0          France  12/09/2011   
541906  12/09/2011 12:50       4.15     12680.0          France  12/09/2011   
541907  12/09/2011 12:50       4.15     12680.0          France  12/09/2011   
541908  12/09/2011 12:50       4.95     12680.0          France  12/09/2011   

        day  month  year  
0         1     12  2010  
1         1     12  2010  
2         1     12  2010  
3         1     12  2010  
4         1     12  2010  
...     ...    ...   ...  
541904    9     12  2011  
541905    9     12  2011  
541906    9     12  2011  
541907    9     12  2011  
541908    9     12  2011  

[68006 rows x 12 columns]

In [35]:
transactionsChristmasNewYear = data_ChristmasNewYear.groupby('InvoiceNo')['StockCode'].apply(list)

In [36]:
!pip install apyori #apriori คนสร้างแพ๊คเกจชอบเอาคำว่าapyไปใส่ เพื่อให้รู้ว่าเป็นแพ๊คเกจของไพทอน

  Preparing metadata (setup.py) ... done
  Created wheel for apyori: filename=apyori-1.1.2-py3-none-any.whl size=5954 sha256=f8a807f71a260f68c62e312ac3deade009c6ddc46ad846f02e3a8aed07483719
  Stored in directory: /root/.cache/pip/wheels/c4/1a/79/20f55c470a50bb3702a8cb7c94d8ada15573538c7f4baebe2d
Successfully built apyori


In [37]:
from apyori import apriori

In [39]:
frequentItemsets = list(apriori(transactionsChristmasNewYear,min_support = 0.05))
#frequentItemsets

In [41]:
min_sup = 0.04

frequentItemsetsC_and_N = list(apriori(transactionsChristmasNewYear,min_support = min_sup))
for i in frequentItemsetsC_and_N:
    if i[0].__len__()>1:
        print('yes')
        print(i)

yes
RelationRecord(items=frozenset({'22086', '22910'}), support=0.047697368421052634, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'22086', '22910'}), confidence=0.047697368421052634, lift=1.0), OrderedStatistic(items_base=frozenset({'22086'}), items_add=frozenset({'22910'}), confidence=0.5160142348754448, lift=7.195794834960331), OrderedStatistic(items_base=frozenset({'22910'}), items_add=frozenset({'22086'}), confidence=0.6651376146788991, lift=7.195794834960332)])


In [44]:
data_ChristmasNewYear[data_ChristmasNewYear['StockCode']=='POST']

InvoiceNo StockCode Description  Quantity       InvoiceDate  UnitPrice  \
45        536370      POST     POSTAGE         3  12/01/2010 08:45       18.0   
386       536403      POST     POSTAGE         1  12/01/2010 11:27       15.0   
1123      536527      POST     POSTAGE         1  12/01/2010 13:04       18.0   
5073      536840      POST     POSTAGE         1  12/02/2010 18:27       18.0   
5258      536852      POST     POSTAGE         1  12/03/2010 09:51       18.0   
...          ...       ...         ...       ...               ...        ...   
541198    581493      POST     POSTAGE         1  12/09/2011 10:10       15.0   
541216    581494      POST     POSTAGE         2  12/09/2011 10:13       18.0   
541730    581570      POST     POSTAGE         1  12/09/2011 11:59       18.0   
541767    581574      POST     POSTAGE         2  12/09/2011 12:09       18.0   
541768    581578      POST     POSTAGE         3  12/09/2011 12:16       18.0   

        CustomerID      Country        date  day  month  year  
45         12583.0       France  12/01/2010    1     12  2010  
386        12791.0  Netherlands  12/01/2010    1     12  2010  
1123       12662.0      Germany  12/01/2010    1     12  2010  
5073       12738.0      Germany  12/02/2010    2     12  2010  
5258       12686.0       France  12/03/2010    3     12  2010  
...            ...          ...         ...  ...    ...   ...  
541198     12423.0      Belgium  12/09/2011    9     12  2011  
541216     12518.0      Germany  12/09/2011    9     12  2011  
541730     12662.0      Germany  12/09/2011    9     12  2011  
541767     12526.0      Germany  12/09/2011    9     12  2011  
541768     12713.0      Germany  12/09/2011    9     12  2011  

[121 rows x 12 columns]

สรุปได้ดังนั้น
* RelationRecord(items=frozenset({'22910', '22086'}), support=0.047697368421052634)
  * ordered_statistics=OrderedStatistic(items_base=frozenset(), items_add=frozenset({'22910', '22086'}), confidence=0.047697368421052634, lift=1.0)
  * OrderedStatistic(items_base=frozenset({'22086'}), items_add=frozenset({'22910'}), confidence=0.5160142348754448, lift=7.195794834960331)
  * OrderedStatistic(items_base=frozenset({'22910'}), items_add=frozenset({'22086'}), confidence=0.6651376146788991, lift=7.195794834960332)


In [43]:
finditem('22910','22086')

{"PAPER CHAIN KIT 50'S CHRISTMAS ", 'PAPER CHAIN KIT VINTAGE CHRISTMAS'}

In [45]:
ChristmasNewYear = data_ChristmasNewYear.groupby(['Country', 'month'])['StockCode'].apply(list)

In [46]:
transactionsChristmasNewYear

InvoiceNo
536365     [85123A, 71053, 84406B, 84029G, 84029E, 22752,...
536366                                        [22633, 22632]
536367     [84879, 22745, 22748, 22749, 22310, 84969, 226...
536368                          [22960, 22913, 22912, 22914]
536369                                               [21756]
                                 ...                        
C581484                                              [23843]
C581490                                       [22178, 23144]
C581499                                                  [M]
C581568                                              [21258]
C581569                                       [84978, 20979]
Name: StockCode, Length: 3040, dtype: object

In [47]:
min_sup =0.3
frequentItemsetsChristmasNewYear = list(apriori(transactionsChristmasNewYear,min_support = min_sup))
for i in frequentItemsetsChristmasNewYear:
    if i[0].__len__()>1:
        print('yes')
        print(i)

In [48]:
finditem('POST', '20750')

{'POSTAGE', 'RED RETROSPOT MINI CASES', nan}

## 2. ของช่วงเวลาในทวีป Europe